### PR TITLE
Initial cmd state

### PR DESCRIPTION
**Pull request by rejmond**
Signed-off-by: Sergey Shlyanin [sergey.shlyanin@xored.com](mailto:sergey.shlyanin@xored.com)

# Description

Clone `cmd-nse-icmp-responder` repository content and rename package to `nse-istio-proxy`.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -14,7 +15,386 @@
...
+	BaseDir               string            `default:"./" desc:"base directory" split_words:"true"`
+	ConnectTo             url.URL           `default:"unix:///var/lib/networkservicemesh/nsm.io.sock" desc:"url to connect to" split_words:"true"`
+	MaxTokenLifetime      time.Duration     `default:"10m" desc:"maximum lifetime of tokens" split_words:"true"`
+	ServiceNames          []string          `default:"istio-proxy-responder" desc:"Name of provided services" split_words:"true"`
+	Payload               string            `default:"ETHERNET" desc:"Name of provided service payload" split_words:"true"`
+	Labels                map[string]string `default:"" desc:"Endpoint labels"`
+	DNSConfigs            dnstools.Decoder  `default:"[]" desc:"DNSConfigs represents array of DNSConfig in json format. See at model definition: https://github.com/networkservicemesh/api/blob/main/pkg/api/networkservice/connectioncontext.pb.go#L426-L435" split_words:"true"`
+	CidrPrefix            []string          `default:"169.254.0.0/16" desc:"List of CIDR Prefix to assign IPv4 and IPv6 addresses from" split_words:"true"`
```
Comment (edwarnicke) [inline main.go:90]: https://github.com/networkservicemesh/api/blob/main/pkg/api/networkservice/connectioncontext.pb.go#L426-L435 will move as main moves, use an absolute reference to a tag or commit for this comment.

File: main.go [hunk 2]
```diff
@@ -14,7 +15,386 @@
...
+	IdleTimeout           time.Duration     `default:"0" desc:"timeout for automatic shutdown when there were no requests for specified time. Set 0 to disable auto-shutdown." split_words:"true"`
+	RegisterService       bool              `default:"true" desc:"if true then registers network service on startup" split_words:"true"`
```
Comment (edwarnicke) [inline main.go:92]: When would this be false?

File: main.go [hunk 3]
```diff
@@ -14,7 +15,386 @@
...
+	if err := debug.Self(); err != nil {
+		log.FromContext(ctx).Infof("%s", err)
+	}
+
+	// enumerating phases
+	log.FromContext(ctx).Infof("there are 6 phases which will be executed followed by a success message:")
+	log.FromContext(ctx).Infof("the phases include:")
+	log.FromContext(ctx).Infof("1: get config from environment")
+	log.FromContext(ctx).Infof("2: retrieve spiffe svid")
+	log.FromContext(ctx).Infof("3: create icmp server ipam")
```
Comment (edwarnicke) [inline main.go:139]: We should fix the text here as its not icmp server.  I'd suggest just changing it to 'create ipam' (and below to 'create nse'

File: main.go [hunk 4]
```diff
@@ -14,7 +15,386 @@
...
+	// Otherwise wait for an error in the background to log and cancel
+	go func(ctx context.Context, errCh <-chan error) {
+		err := <-errCh
+		log.FromContext(ctx).Error(err)
+		cancel()
+	}(ctx, errCh)
+}
+
+// newPolicyRoutesGetter - watches the config file and dynamically updates policy routes.
+func newPolicyRoutesGetter(ctx context.Context, configPath string) *policyRoutesGetter {
```
Comment (edwarnicke) [inline main.go:347]: This whole function is probably unneeded in this NSE.

File: main.go [hunk 5]
```diff
@@ -14,7 +15,386 @@
...
+			case update := <-updateCh:
+				updatePrefixes(update)
+			}
+		}
+	}()
+
+	return p
+}
+
+func (p *policyRoutesGetter) Get() []*networkservice.PolicyRoute {
```
Comment (edwarnicke) [inline main.go:380]: This whole function is probably unneeded in this NSE

**Comment by rejmond** [comment]
@edwarnicke Thank you for your comments, I will check everything.

**Review by denis-tingaikin — APPROVED** [review]
@rejmond Please resolve @edwarnicke comment and this can be merged. Thanks.

File: main.go (hunk 2 above)
Comment (rejmond) [reply main.go]: Done. Removed this option.

File: main.go [hunk 6]
```diff
@@ -14,7 +15,386 @@
...
+)
+
+// Config holds configuration parameters from environment variables
+type Config struct {
+	Name                  string            `default:"icmp-server" desc:"Name of ICMP Server"`
+	BaseDir               string            `default:"./" desc:"base directory" split_words:"true"`
+	ConnectTo             url.URL           `default:"unix:///var/lib/networkservicemesh/nsm.io.sock" desc:"url to connect to" split_words:"true"`
+	MaxTokenLifetime      time.Duration     `default:"10m" desc:"maximum lifetime of tokens" split_words:"true"`
+	ServiceNames          []string          `default:"istio-proxy-responder" desc:"Name of provided services" split_words:"true"`
+	Payload               string            `default:"ETHERNET" desc:"Name of provided service payload" split_words:"true"`
```
[the remark this answers is not in the document]
Comment (rejmond) [reply main.go]: Done. IP contants is used now in the registry client.

File: main.go (hunk 4 above)
Comment (rejmond) [reply main.go]: Done.

File: main.go (hunk 5 above)
Comment (rejmond) [reply main.go]: Done.

**Comment by denis-tingaikin** [comment]
@edwarnicke Let us know if you still have any comments on this then we'll resolve them in separate PR.

**Pull request merged**